### PR TITLE
chore: Prevent test suite artifact creation in work directory

### LINKED ISSUE #3520: temporary directories not cleaned up after tests
`type_defect`   |    by jennifer@painless-security.com

___

Two temporary directories are being left around after running the tests:
```
$ ls -ld tmp-*
drwxr-xr-x 55 root root 4096 Jan 24 17:26 tmp-nomcom-public-keys-dir
drwxr-xr-x  5 root root 4096 Jan 24 17:06 tmp-storage-meetinghosttests-dir
```
These should be cleaned up when the tests are done with them.

___
_Issue migrated from trac:3520 at 2022-03-04 09:25:31 +0000_

### LINKED ISSUE #4020: Make test suite generate files outside of the source tree
### Description

The tests seem to generate directories in the source tree names `tmp-*` and they also seem to generate files under `test/media` and possibly `test` in general.

It would be nice if that as reworked so all that happened under `/tmp`.

### Code of Conduct

- [X] I agree to follow the [IETF's Code of Conduct](https://github.com/ietf-tools/.github/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by larseggert**
Also remove a few other stale test assets while I'm here.

Fixes #4020
Fixes #3520

### CODE REVIEW

File: dev/tests/settings_local.py
Comment (rjsparks) [inline dev/tests/settings_local.py:20]: This should go into `/assets/ietfdata/doc/draft/archive` to match the tree currently in production at ietfdata (and things that are currently in www6 will move into that tree as we move all of this into a blob-store).

Similarly the repository should be at `/assets/ietfdata/doc/draft/repository`

and other document types are simpler:

```
./doc/bofreq
./doc/charter
./doc/conflict-review
./doc/review
./doc/slides # This one is strange - it's edu + iab ws slides at the moment, not meeting materials slides
./doc/status-change
./doc/statement
```

File: dev/tests/settings_local.py
Comment (rjsparks) [inline dev/tests/settings_local.py:21]: in the long run, staging needs to be reworked anyhow to be in db, not on disk. I'd like to squeeze www6s out of the dev container configuration, but until we pull everything out of the proceedings dir (where the meeting materials are currently living), we can keep /assets/www6s to reflect the production reality.

**Review by rjsparks — CHANGES_REQUESTED** [review]

File: dev/tests/settings_local.py
Comment (larseggert) [reply dev/tests/settings_local.py:20]: "This should go" - which this? `IDSUBMIT_REPOSITORY_PATH` or `IDSUBMIT_STAGING_PATH`?

Comment (rjsparks) [reply dev/tests/settings_local.py:20]: `IDSUBMIT_REPOSITORY_PATH`

Thanks for asking - my direction above was a little off.
In production right now:
https://github.com/ietf-tools/datatracker/blob/a02812bf7c7177471af18141bcf6cae6375d9ee8/ietf/settings.py#L674
and 
https://github.com/ietf-tools/datatracker/blob/a02812bf7c7177471af18141bcf6cae6375d9ee8/ietf/settings.py#L799

We should mirror that in the dev instance (it's important that the submisssion tool post draft into the repository, not the archive).

That is, we should have this in the dev instance:
`IDSUBMIT_REPOSITORY_PATH = /assets/ietfdata/doc/draft/repository`

**Comment by larseggert** [comment]
@rjsparks 
>CRITICALS:
?: (datatracker.E0006) A directory used by the I-D submission tool does not
exist at the path given in the settings file.  The setting is:
    IDSUBMIT_REPOSITORY_PATH = /assets/ietfdata/doc/draft/repository

**Comment by larseggert** [comment]
There are a ton of directories that are being created in `docker/scripts/app-create-dirs.sh` - it's unclear to me how many of them are actually needed to run a dev instance and the CI.

**Comment by larseggert** [comment]
@rjsparks I think I made all the changes you requested?

**Comment by rjsparks** [comment]
Yes, I'm holding this for more work post the feat/rfc merge.

**Comment by rjsparks** [comment]
This didn't get returned to post feat/rfc. We should review it fresh to see if it's still the right thing to do.
cc: @jennifer-richards

**Comment by jennifer-richards** [comment]
I left a couple comments - cleaning up after the tests, even in `/tmp` or wherever those paths end up, would be nice but also tricky. If the tests pass against modern main, I would be inclined to accept this as an improvement and treat those future work.

File: ietf/settings_test.py
Comment (jennifer-richards) [reply ietf/settings_test.py]: Fixed in e9bc0ae29ecedf9eb1caedf3e9b7c29ff0f6a21a

**Comment by jennifer-richards** [comment]
Hopefully not stepping on @larseggert's toes, but just pushed changes that address my concerns. I think it's ready for @rjsparks to re-review?

**Comment by larseggert** [comment]
No toes in the area :-) I won't get to this any time soon, so please do what you need to!

File: ietf/person/factories.py
Comment (rjsparks) [inline ietf/person/factories.py:108]: remind me why we're not just making a copy of the profile-default.jpg again?
(I'm sensitive to the PIL import - it's large)

File: media/photo/profile-default.jpg
Comment (rjsparks) [inline media/photo/profile-default.jpg:1]: Doesn't the non-test code _use_ this for displaying the profile page when no picture is available?

**Review by rjsparks — COMMENTED** [review]
I am a little nervous at the "clean up on exit" vs "clean up as you go" approach as it adds to the recovery for the developer if the process goes away via `kill -9` or the like, but am willing to try it.

Just one question about why we're trading a file that I _think_ we should be keeping for PIL.

**Comment by jennifer-richards** [comment]
> I am a little nervous at the "clean up on exit" vs "clean up as you go" approach as it adds to the recovery for the developer if the process goes away via `kill -9` or the like, but am willing to try it.

I'm sympathetic. The `kill -9` doesn't worry me terribly because it's rare and the impact is just a couple of directories in `/tmp` left over and `atexit` is robust against most normal exits.

I'd _much_ rather create the temp dir separately for each test to maintain isolation. Before adding the `atexit` business I tried using the temp directory mechanism we have in our `TestCase` class and it kinda sorta worked, but ran into bad interactions with the test startup sequence. It definitely interacted with our checks for the directories existing, but IIRC the statics serving for tests was also broken by this because it's only set up once for the entire test run.

File: media/photo/profile-default.jpg
Comment (jennifer-richards) [reply media/photo/profile-default.jpg:1]: (and for the thumbnail, we use an icon - it probably resembles the profile-default image so maybe that's what you're thinking of?)

https://github.com/ietf-tools/datatracker/blob/a6377997953ced70c78c8c3a3a33ee17a31dc8f3/ietf/templates/person/photo.html#L13-L15

File: ietf/person/factories.py
Comment (jennifer-richards) [reply ietf/person/factories.py:108]: I think it's just avoiding a file that (as far as I can tell) exists only for the tests.

I don't have an opinion one way or the other, but maybe there's a lighter weight library for generating an empty image for testing.

Comment (jennifer-richards) [reply ietf/person/factories.py:108]: There's PyPNG - much smaller it seems, but requires less clearly intentioned code to generate a blank image and is somewhat slower on a per-image basis. The import of PIL on my system looks to be a few tens of ms - and I think we probably only hit that once per test run. (Though I haven't actually done a profile to see if the import is being exercised more than once; not sure how to do that)

**Review by rjsparks — APPROVED** [review]

File: ietf/person/factories.py
Comment (rjsparks) [reply ietf/person/factories.py:108]: I suspect there's a valid, small, literal we can use as a PNG, but not going to slow this down over it.
I think we should, however, spend time scrubbing our full dependency set and seeing if we can make it smaller, and PIL is one of the things I'm wondering if we can scrub completely away.

Comment (jennifer-richards) [reply ietf/person/factories.py:108]: We use PIL for resizing images in `Person.photo`. If we do away with that (mis?)feature, then we can probably ditch even needing to fake an image. 

Checked and PyPNG is a < 40kb python file. It can read and write PNGs, so if we _do_ need to create one for some reason, it's a pretty minimal bit of baggage.

**Pull request merged**
